### PR TITLE
Add better printing in reference interpreter, add tolerance to check almost eq

### PR DESCRIPTION
**Pull request by GleasonK**
Background: I was running interpreter testcases and wanted some syntax I could copy-paste into python. Now tensors in the interpreter print in a python-array-compatible syntax, wrapped in a `tensor{...}` descriptor.

```
$ cat /tmp/t.mlir 
func.func @main() -> tensor<2x2xi8> {
  %matrix = stablehlo.constant dense<[[1, 2], [3, 4]]> : tensor<2x2xi8>
  func.return %matrix : tensor<2x2xi8>
}

$ stablehlo-translate --interpret /tmp/t.mlir 
tensor<2x2xi8> {
  [
    [1, 2],
    [3, 4]
  ]
}
```

Also as a part of this work I needed to modify some testcase tolerances.

### CODE REVIEW

File: stablehlo/tests/interpret/check.mlir [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+func.func @check_almost_eq_const_tolerance() {
+  %operand = stablehlo.constant dense<5.0000> : tensor<f64>
+  check.expect_almost_eq_const %operand, dense<5.1> : tensor<f64> {tolerance = 0.1 : f64}
+  func.return
+}
+
+// -----
+
+func.func @check_close() {
```
Comment (sdasgup3) [inline stablehlo/tests/interpret/check.mlir:55]: Is this test added just for covering all kinds of check ops.

File: stablehlo/reference/Tensor.cpp [hunk 2]
```diff
@@ -370,13 +373,53 @@ IndexSpaceIterator Tensor::index_begin() const {
 
 IndexSpaceIterator Tensor::index_end() const { return getShape().index_end(); }
 
-void Tensor::print(raw_ostream &os) const {
-  getType().print(os);
-  os << " {\n";
+namespace {
+
+void printNewline(llvm::raw_ostream &os, int64_t n) {
```
Comment (sdasgup3) [inline stablehlo/reference/Tensor.cpp:378]: `printIndentation` may be ... as this is doing more than printing newlines.

File: stablehlo/reference/Tensor.cpp [hunk 3]
```diff
@@ -370,13 +373,53 @@ IndexSpaceIterator Tensor::index_begin() const {
...
+  os << '\n';
+  for (int64_t i = 0; i < n; ++i) os << "  ";
+}
+
+bool isLast(const Index &idx, const Sizes &shape) {
```
Comment (sdasgup3) [inline stablehlo/reference/Tensor.cpp:383]: `isEndOfIterationSpace`?

File: stablehlo/reference/Tensor.cpp [hunk 4]
```diff
@@ -370,13 +373,53 @@ IndexSpaceIterator Tensor::index_begin() const {
...
+  // {0, 3} vs {0, 4, 9} ==> true
```
Comment (sdasgup3) [inline stablehlo/reference/Tensor.cpp:384]: Can you elaborate on this comment? Or maybe the name is self descriptive?

**Review by sdasgup3 — APPROVED** [review]
lgtm except for a few trivial comment.

File: stablehlo/tests/interpret/check.mlir (hunk 1 above)
Comment (GleasonK) [reply stablehlo/tests/interpret/check.mlir:55]: All except the probe one which is tested elsewhere

File: stablehlo/reference/Tensor.cpp (hunk 2 above)
Comment (GleasonK) [reply stablehlo/reference/Tensor.cpp:378]: went with `printNewlineIndent`

File: stablehlo/tests/interpret/printing.mlir [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+  // CHECK:      tensor<2xi1> {
+  // CHECK-NEXT:   [true, false]
+  // CHECK-NEXT: }
+  %bool = stablehlo.constant dense<[true, false]> : tensor<2xi1>
+
+  // CHECK:      tensor<0xui8> {
+  // CHECK-NEXT:   []
+  // CHECK-NEXT: }
+  %empty = stablehlo.constant dense<> : tensor<0xui8>
```
[the remark this answers is not in the document]
Comment (GleasonK) [reply stablehlo/tests/interpret/printing.mlir]: Added.

```
  // CHECK:      tensor<0xui8> {
  // CHECK-NEXT:   []
  // CHECK-NEXT: }
  %empty_scalar = stablehlo.constant dense<> : tensor<0xui8>

  // CHECK:      tensor<0x2xi8> {
  // CHECK-NEXT:   []
  // CHECK-NEXT: }
  %empty_matrix = stablehlo.constant dense<> : tensor<0x2xi8>

  // CHECK:      tensor<2x0xi8> {
  // CHECK-NEXT:   [
  // CHECK-NEXT:     [],
  // CHECK-NEXT:     []
  // CHECK-NEXT:   ]
  // CHECK-NEXT: }
  %empty_matrix_col = stablehlo.constant dense<> : tensor<2x0xi8>
```

I don't have strong opinions on the empty tensors, they're not common enough to warrant the extra code tech-debt to print them in a less ambiguous way. Also with the `tensor` type, I don't think there's ambiguity, just won't be copy-pastable to python

File: stablehlo/reference/Tensor.cpp [hunk 6]
```diff
@@ -16,13 +16,16 @@ limitations under the License.
...
 
 #include "llvm/ADT/APFloat.h"
 #include "llvm/ADT/APInt.h"
 #include "llvm/Support/Error.h"
+#include "llvm/Support/raw_ostream.h"
 #include "mlir/IR/BuiltinAttributes.h"
 #include "mlir/IR/BuiltinTypes.h"
 #include "mlir/Support/DebugStringHelper.h"
+#include "stablehlo/reference/Element.h"
 #include "stablehlo/reference/Errors.h"
```
[the remark this answers is not in the document]
Comment (GleasonK) [reply stablehlo/reference/Tensor.cpp:29]: `StablehloReferenceElement` already included?

[the remark this answers is not in the document]
Comment (GleasonK) [reply stablehlo/reference/Tensor.cpp:29]: Original print was using `os << get(*it)` which returns element so we probably were just missing the explicit include. Also raw_ostream comes in MLIRIR by the looks of it

[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/reference/Tensor.cpp:29]: Ah sorry should've clarified the file name; it looks like "Errors.h" isn't explicitly included.

**Review by ghpvnist — APPROVED** [review]
LGTM once the remaining comments are addressed. Love the new format!

**Pull request merged**